ADR/nnnn-task-versioning.md

```diff
@@ -0,0 +1,361 @@
+- The latest revision of each `0.x` version is trusted, until Task maintainers decide
+  this particular version should enter the expiry period.
+- When the `0.x` version tag moves to a new revision, the older revision (previously
+  tagged as `0.x`) begins its expiry period.
+- *The same would likely apply to `1.x`, `2.x` etc. We don't know, no Konflux Task
+  has made it that far.*
+
+This mechanism is rather clunky and not very flexible. And, for reasons explained
+in [ADR NN](ADR/tbd.md), the mechanism will not work after the build-definitions
+decentralization.
+
+A more flexible solution would allow the policy author to allow/disallow specific
+version ranges. Such a solution cannot work if dozens of different versions of the
+same Task share the same version number.
+
+## Decision
+
+### Mark each meaningful change with a version update
+
+For Tasks at version `0.x`, you may continue to mark breaking changes with minor
+version updates and start marking non-breaking changes with patch version updates.
+Do aim to bring the Task to version `1.0`.
+
+For Tasks at version `1.0` or greater, follow [Semantic Versioning](https://semver.org/).
+
+This ADR leaves some freedom to each Task maintainer team to decide how to ensure
+changes get versioned. The *suggested* approach would be to version manually
+and set up CI in a way that encourages/enforces versioning. See Appendix 1:
+[The technicals of Task versioning](#the-technicals-of-task-versioning).
+
+### Don't require Task repositories to follow Tekton Catalog layout
+
+The [Tekton Catalog layout], which places each version of a Task in a separate
+`task/${task_name}/${version}/` directory, causes far too much overhead when
+updating Task versions. If we're going to require version updates for each meaningful
+change, we cannot ask Task maintainers to follow this layout. See Appendix 1:
+[The technicals of Task versioning](#the-technicals-of-task-versioning) for more details.
+
+### Keep a changelog
+
+Version numbers on their own are nice, but without an easy way to see what changed
+in each version, they're not that useful.
+
+Document the meaningful changes in each version in a CHANGELOG.md file.
+
+### MintMaker: don't assume each x.y version has a MIGRATION.md
+
+MintMaker links MIGRATION.md documents in the PRs it sends to Konflux users (see
+the example [above](#users-get-opaque-updates)). This feature assumes that each
+`major.minor` version comes with a MIGRATION.md document, that this document is
+at `/task/${task_name}/${version}/MIGRATION.md` in the source repository and that
+the source repository is <https://github.com/konflux-ci/build-definitions>. The
+feature will not work after decentralization.
+
+<!-- TODO: How will MintMaker figure out a link to the changelog? -->
+MintMaker will stop trying to link MIGRATION.md files and instead link the changelog.
+In case of breaking changes, the changelog should explain how the user can migrate.
```

Comment:
@konflux-ci/mintmaker-maintainers I would appreciate your help with this.

Did you have some plans for how the MIGRATION.md links would keep working after decentralization? Or any ideas how we could make CHANGELOG.md discovery work?

Reply:
I know that Renovate also has some support for discovering CHANGELOG.md files in the source repo (this ADR proposes we should use those): https://docs.renovatebot.com/key-concepts/changelogs/

But I imagine this looks for CHANGELOG.md in the repo root or some other known locations and also wouldn't know where to find the correct file for a Tekton task.

> We can potentially implement a new source for the releases in our downstream fork of Renovate.

Yeah, that might be the way. I'm thinking we would include a link to the CHANGELOG.md as an annotation on the task bundle (similar to [this one](https://github.com/konflux-ci/build-definitions/blob/9989c73a5246e6b52f53ff4dff2ffc678c8d1630/task/tkn-bundle/0.2/tkn-bundle.yaml#L404)) and MintMaker would use that